Google.Api.Gax.Grpc/GoogleCredentialExtensions.cs

```diff
@@ -0,0 +1,73 @@
+﻿/*
+ * Copyright 2024 Google LLC
+ * Use of this source code is governed by a BSD-style
+ * license that can be found in the LICENSE file or at
+ * https://developers.google.com/open-source/licenses/bsd
+ */
+
+using Google.Apis.Auth.OAuth2;
+using Grpc.Auth;
+using Grpc.Core;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Api.Gax.Grpc
+{
+    /// <summary>
+    /// Extension methods for Google credential universe domain validation.
+    /// </summary>
+    public static class GoogleCredentialExtensions
+    {
+        /// <summary>
+        /// Returns a channel credential based on <paramref name="googleCredential"/>, that will validate its own universe domain
+        /// against <paramref name="universeDomain"/>.
+        /// </summary>
+        /// <param name="googleCredential">The Google credential to build the channel credentials from. Must not be null.</param>
+        /// <param name="universeDomain">The universe domain to validate against. Must not be null.
+        /// <see cref="GoogleCredential.GetUniverseDomainAsync(CancellationToken)"/> should result in the same value as this.</param>
+        public static ChannelCredentials ToChannelCredentials(this GoogleCredential googleCredential, string universeDomain) =>
+            new GoogleCredentialWithUniverseDomainValidation(googleCredential, universeDomain).ToChannelCredentials();
+
+        private class GoogleCredentialWithUniverseDomainValidation : ITokenAccessWithHeaders
+        {
+            private readonly ITokenAccessWithHeaders _underlying;
+            private readonly string _universeDomain;
+            private readonly Lazy<Task<bool>> _universeDomainsMatchCache;
```

Comment:
Why not just `Lazy<Task>` out of interest?

Reply:
Oh, but when changing, I've needed to add a WithCancellationToken for just Task. Which is fine.